### ISSUE TITLE
Opam fails to install base64(and usbmux)

### ISSUE THREAD

**Issue by Canorus** [onlinemediagroup/ocaml-usbmux#9]
Here's what's printed out.

```
Canors-Mac-mini:~ Canor$ opam install usbmux
The following actions will be performed:
  ∗  install ocamlmod        0.0.8            [required by oasis]
  ∗  install biniou          1.0.12           [required by yojson]
  ∗  install stringext       1.4.2            [required by usbmux]
  ∗  install lwt             2.5.2            [required by usbmux]
  ∗  install base64          2.0.0            [required by cohttp]
  ∗  install magic-mime      1.0.0            [required by cohttp]
  ∗  install ppx_deriving    3.3              [required by ppx_type_conv]
  ∗  install ocplib-endian   1.0              [required by cstruct]
  ∗  install re              1.6.1            [required by cohttp]
  ∗  install ppx_core        113.33.01+4.03   [required by ppx_fields_conv]
  ∗  install sexplib         113.33.00+4.03   [required by cohttp]
  ∗  install fieldslib       113.24.00        [required by cohttp]
  ∗  install ocamlify        0.0.1            [required by oasis]
  ∗  install yojson          1.3.2            [required by usbmux]
  ∗  install markup          0.7.2            [required by lambdasoup]
  ∗  install ppx_optcomp     113.33.00+4.03   [required by conduit]
  ∗  install cstruct         2.3.0            [required by conduit]
  ∗  install oasis           0.4.7            [required by usbmux]
  ∗  install lambdasoup      0.6              [required by plist]
  ∗  install ppx_driver      113.33.01+4.03   [required by conduit]
  ∗  install plist           0.1              [required by usbmux]
  ∗  install ppx_type_conv   113.33.01+4.03   [required by ppx_fields_conv]
  ∗  install ppx_sexp_conv   113.33.01+4.03   [required by cohttp]
  ∗  install ppx_fields_conv 113.33.00+4.03   [required by cohttp]
  ∗  install uri             1.9.2            [required by cohttp]
  ∗  install ipaddr          2.7.0            [required by conduit]
  ∗  install conduit         0.12.0           [required by cohttp]
  ∗  install cohttp          0.21.0           [required by usbmux]
  ∗  install usbmux          1.1.1         
===== ∗  29 =====
Do you want to continue ? [Y/n] y

=-=- Gathering sources =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[base64] Archive in cache
[biniou] Archive in cache
[cohttp] Archive in cache
[conduit] Archive in cache
[cstruct] Archive in cache
[fieldslib] Archive in cache
[ipaddr] Archive in cache
[lambdasoup] Archive in cache
[lwt] Archive in cache
[magic-mime] Archive in cache
[markup] Archive in cache
[oasis] Archive in cache
[ocamlify] Archive in cache
[ocamlmod] Archive in cache
[ocplib-endian] Archive in cache
[plist] Archive in cache
[ppx_core] Archive in cache
[ppx_deriving] Archive in cache
[ppx_driver] Archive in cache
[ppx_fields_conv] Archive in cache
[ppx_optcomp] Archive in cache
[ppx_sexp_conv] Archive in cache
[ppx_type_conv] Archive in cache
[re] Archive in cache
[sexplib] Archive in cache
[stringext] Archive in cache
[uri] Archive in cache
[usbmux] Archive in cache
[yojson] Archive in cache

=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

#=== ERROR while installing base64.2.0.0 ======================================#
# opam-version    1.2.2
# os              darwin
Extracting the archive will overwrite /Users/Canor 1/.opam/system/build/base64.2.0.0.


=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
The following actions were aborted
  ∗  install biniou          1.0.12        
  ∗  install cohttp          0.21.0        
  ∗  install conduit         0.12.0        
  ∗  install cstruct         2.3.0         
  ∗  install fieldslib       113.24.00     
  ∗  install ipaddr          2.7.0         
  ∗  install lambdasoup      0.6           
  ∗  install lwt             2.5.2         
  ∗  install magic-mime      1.0.0         
  ∗  install markup          0.7.2         
  ∗  install oasis           0.4.7         
  ∗  install ocamlify        0.0.1         
  ∗  install ocamlmod        0.0.8         
  ∗  install ocplib-endian   1.0           
  ∗  install plist           0.1           
  ∗  install ppx_core        113.33.01+4.03
  ∗  install ppx_deriving    3.3           
  ∗  install ppx_driver      113.33.01+4.03
  ∗  install ppx_fields_conv 113.33.00+4.03
  ∗  install ppx_optcomp     113.33.00+4.03
  ∗  install ppx_sexp_conv   113.33.01+4.03
  ∗  install ppx_type_conv   113.33.01+4.03
  ∗  install re              1.6.1         
  ∗  install sexplib         113.33.00+4.03
  ∗  install stringext       1.4.2         
  ∗  install uri             1.9.2         
  ∗  install usbmux          1.1.1         
  ∗  install yojson          1.3.2         
The following actions failed
  ∗  install base64 2.0.0
No changes have been performed
```

it's OSX 10.11.6, opam compiler 4.03.0 installed via homebrew, however Terminal prints 

```
The environment is not in sync with the current switch.
          You should run: eval `opam config env`
```

when I type `opam switch`.

**Comment by fxfactorial** [comment]
Thanks for the bug report! 

First try doing what the switch warning says: `eval `opam config env`` , then try `opam install usbmux` again.

**Comment by Canorus** [comment]
Thanks for quick replay.

It changed nothing. Just in case it would help you figure out what I did wrong I append the result of `opam config env`

```
Canors-Mac-mini:~ Canor$ opam config env
CAML_LD_LIBRARY_PATH="/Users/Canor 1/.opam/system/lib/stublibs:/usr/local/lib/ocaml/stublibs"; export CAML_LD_LIBRARY_PATH;
OPAMUTF8MSGS="1"; export OPAMUTF8MSGS;
MANPATH="/Users/Canor 1/.opam/system/man:"; export MANPATH;
PERL5LIB="/Users/Canor 1/.opam/system/lib/perl5"; export PERL5LIB;
OCAML_TOPLEVEL_PATH="/Users/Canor 1/.opam/system/lib/toplevel"; export OCAML_TOPLEVEL_PATH;
PATH="/Users/Canor 1/.opam/system/bin:/usr/local/sbin:/usr/local/Cellar/node/6.3.1/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin"; export PATH;
```

**Comment by fxfactorial** [comment]
This is odd, what is the full error message if you try installing `opam install base64` ? Perhaps try that first, then `opam install usbmux`

**Comment by Canorus** [comment]
This is the full message

```
Canors-Mac-mini:~ Canor$ opam install base64
The following actions will be performed:
  ∗  install base64 2.0.0

=-=- Gathering sources =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[base64] Archive in cache

=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

#=== ERROR while installing base64.2.0.0 ======================================#
# opam-version    1.2.2
# os              darwin
Extracting the archive will overwrite /Users/Canor 1/.opam/system/build/base64.2.0.0.


=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
The following actions failed
  ∗  install base64 2.0.0
No changes have been performed
```

When I tried to install it by manually deleting base64.2.0.0 dir, all related dependency packages seems to fail to install.

**Comment by fxfactorial** [comment]
try this order: 

```
opam switch idevices -A 4.02.3
eval $(opam config env)
opam install usbmux -y
```

**Comment by Canorus** [comment]
after I run first line I got this error message:

```
Canors-Mac-mini:~ Canor$ opam switch idevices -A 4.02.3

=-=- Installing compiler 4.02.3 -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[default.comp] http://caml.inria.fr/pub/distrib/ocaml-4.02/ocaml-4.02.3.tar.gz downloaded
Now compiling OCaml. This may take a while, please bear with us...
[ERROR] Compiler build failed at "make install":
        # opam-version 1.2.2
        # os           darwin
        # command      make install
        # path         /Users/Canor 1/.opam/idevices/build/ocaml
        # exit-code    2
        # env-file     /Users/Canor 1/.opam/log/log-19906-ad8886.env
        # stdout-file  /Users/Canor 1/.opam/log/log-19906-ad8886.out
        # stderr-file  /Users/Canor 1/.opam/log/log-19906-ad8886.err
        ### stdout ###
        # if test -d /Users/Canor 1/.opam/idevices/bin; then : ; \
        #     else mkdir -p /Users/Canor 1/.opam/idevices/bin; fi
        ### stderr ###
        # /bin/sh: line 0: test: /Users/Canor: binary operator expected
        # mkdir: /Users/Canor: Permission denied
        # make: *** [install] Error 1
```

and others went same, failed to install base64.

**Comment by fxfactorial** [comment]
Okay, your permissions look messed up. Did you install brew as sudo? Or opam? Or using one account from another. Why do you have different users, one is `Canor 1`, another is `Canor`

**Comment by Canorus** [comment]
no, none of both was installed as sudo. Should I run permission check?

**Comment by fxfactorial** [comment]
The different users looks weird. `/Users/Canor 1/.opam/` and `/Users/Canor:`

**Comment by Canorus** [comment]
maybe… that's because I restored my mac mini from my macbook and then changed settings, but couldn't change the main HDD name?

![](http://d.pr/i/18RyO+)

could that be the cause?

**Comment by fxfactorial** [comment]
@Canorus I'm kinda suspicious about that, yea.

**Comment by Canorus** [comment]
maybe it's a good time for me to change the name of home directory. I'll post here if it doesn't fix the problem. Thanks!!

**Comment by Canorus** [comment]
I've changed the name and here's the result. It was performed after `eval $(opam config env)`

```
Canors-Mac-mini:~ Canor$ opam install usbmux
The following actions will be performed:
  ∗  install ocamlmod        0.0.8            [required by oasis]
  ∗  install biniou          1.0.12           [required by yojson]
  ∗  install stringext       1.4.2            [required by usbmux]
  ∗  install lwt             2.5.2            [required by usbmux]
  ∗  install base64          2.0.0            [required by cohttp]
  ∗  install magic-mime      1.0.0            [required by cohttp]
  ∗  install ppx_deriving    3.3              [required by ppx_type_conv]
  ∗  install ocplib-endian   1.0              [required by cstruct]
  ∗  install re              1.6.1            [required by cohttp]
  ∗  install ppx_core        113.33.01+4.03   [required by ppx_fields_conv]
  ∗  install sexplib         113.33.00+4.03   [required by cohttp]
  ∗  install fieldslib       113.24.00        [required by cohttp]
  ∗  install ocamlify        0.0.1            [required by oasis]
  ∗  install yojson          1.3.2            [required by usbmux]
  ∗  install markup          0.7.2            [required by lambdasoup]
  ∗  install ppx_optcomp     113.33.00+4.03   [required by conduit]
  ∗  install cstruct         2.3.0            [required by conduit]
  ∗  install oasis           0.4.7            [required by usbmux]
  ∗  install lambdasoup      0.6              [required by plist]
  ∗  install ppx_driver      113.33.01+4.03   [required by conduit]
  ∗  install plist           0.1              [required by usbmux]
  ∗  install ppx_type_conv   113.33.01+4.03   [required by ppx_fields_conv]
  ∗  install ppx_sexp_conv   113.33.01+4.03   [required by cohttp]
  ∗  install ppx_fields_conv 113.33.00+4.03   [required by cohttp]
  ∗  install uri             1.9.2            [required by cohttp]
  ∗  install ipaddr          2.7.0            [required by conduit]
  ∗  install conduit         0.12.0           [required by cohttp]
  ∗  install cohttp          0.21.0           [required by usbmux]
  ∗  install usbmux          1.1.1         
===== ∗  29 =====
Do you want to continue ? [Y/n] y

=-=- Gathering sources =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[ERROR] The compilation of fieldslib failed at "./configure --prefix
        /Users/Canor 1/.opam/system".
[ERROR] The compilation of lwt failed at "./configure --prefix /Users/Canor
        1/.opam/system --disable-libev --disable-camlp4 --disable-react
        --disable-ssl --enable-unix --enable-preemptive --disable-glib
        --enable-ppx".
[ERROR] The compilation of magic-mime failed at "./configure
        --prefix=/Users/Canor 1/.opam/system".
[ERROR] The compilation of biniou failed at "make".
[ERROR] The compilation of base64 failed at "ocaml setup.ml -configure --prefix
        /Users/Canor 1/.opam/system".
[ERROR] The compilation of ocamlmod failed at "ocaml setup.ml -configure
        --prefix /Users/Canor 1/.opam/system".
[ERROR] The compilation of ocamlify failed at "ocaml setup.ml -configure
        --prefix /Users/Canor 1/.opam/system".
[ERROR] The compilation of ocplib-endian failed at "ocaml setup.ml -configure
        --disable-debug --prefix /Users/Canor 1/.opam/system".
[ERROR] The compilation of ppx_core failed at "./configure --prefix
        /Users/Canor 1/.opam/system".

#=== ERROR while installing ppx_deriving.3.3 ==================================#
OpamSystem.File_not_found("/Users/Canor/.opam/system/build/ppx_deriving.3.3/pkg/META.in")


=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
The following actions were aborted
  ∗  install cohttp          0.21.0        
  ∗  install conduit         0.12.0        
  ∗  install cstruct         2.3.0         
  ∗  install ipaddr          2.7.0         
  ∗  install oasis           0.4.7         
  ∗  install ocamlify        0.0.1         
  ∗  install plist           0.1           
  ∗  install ppx_core        113.33.01+4.03
  ∗  install ppx_driver      113.33.01+4.03
  ∗  install ppx_fields_conv 113.33.00+4.03
  ∗  install ppx_optcomp     113.33.00+4.03
  ∗  install ppx_sexp_conv   113.33.01+4.03
  ∗  install ppx_type_conv   113.33.01+4.03
  ∗  install re              1.6.1         
  ∗  install sexplib         113.33.00+4.03
  ∗  install stringext       1.4.2         
  ∗  install uri             1.9.2         
  ∗  install usbmux          1.1.1         
  ∗  install yojson          1.3.2         
The following actions failed
  ∗  install ocplib-endian 1.0
  ∗  install ppx_deriving  3.3
The following changes have been performed
  ∗  install base64     2.0.0    
  ∗  install biniou     1.0.12   
  ∗  install fieldslib  113.24.00
  ∗  install lambdasoup 0.6      
  ∗  install lwt        2.5.2    
  ∗  install magic-mime 1.0.0    
  ∗  install markup     0.7.2    
  ∗  install ocamlmod   0.0.8  
```

I think I'll need to change the path but I can't figure out how to. Thanks.

**Comment by fxfactorial** [comment]
Try an `opam update` first. ppx 3.3 sounds old, maybe. Something still looks messed up from the paths. 

Perhaps a hack/brute force solution is getting rid of opam altogether, making sure that `~/.opam` is gone, then trying again from scratch.

**Comment by Canorus** [comment]
Finally it worked… So basically the inconsistency between home directory name and username was the cause of this problem.

And oddly enough installing usbmux automatically installed ppx 3.3. When I ran `opam upgrade --verbose` it went like below:

```
Canors-Mac-mini:~ Canor$ opam upgrade --verbose
Everything as up-to-date as possible.
The following newer versions couldn't be installed:
  - base-bytes.legacy is not available because it requires OCaml < 4.02.0.
  - fieldslib.113.33.03 is not available because your system doesn't comply with ocaml-version =
    "4.02.3".
  - ppx_core.113.33.03 is not available because your system doesn't comply with ocaml-version =
    "4.02.3".
  - ppx_driver.113.33.03 is not available because your system doesn't comply with ocaml-version =
    "4.02.3".
  - ppx_fields_conv.113.33.03 is not available because your system doesn't comply with
ocaml-version =
    "4.02.3".
  - ppx_optcomp.113.33.03 is not available because your system doesn't comply with ocaml-version =
    "4.02.3".
  - ppx_sexp_conv.113.33.03 is not available because your system doesn't comply with ocaml-version
=
    "4.02.3".
  - ppx_type_conv.113.33.03 is not available because your system doesn't comply with ocaml-version
=
    "4.02.3".
  - sexplib.113.33.03 is not available because your system doesn't comply with ocaml-version =
    "4.02.3".
The following would require downgrades or uninstalls, but you may upgrade them explicitly:
  - ppx_deriving.4.0
  - usbmux.1.3.2
```

so perhaps I should have used compiler 4.02.3 instead of 4.03.0?
Anyways thank you so much for help :D

**Comment by fxfactorial** [comment]
yay! Yea, I think if you do `opam update`, It will be forced to be with just `4.02.3` and lower cause of some upstream packages messing up. 

Glad you're getting usage of it! I have more plans to remove the dependency on `usbmuxd` by adding its functionality into this program itself.
